Product/Samples/Clock/Clock/GPIOButtonInputProvider.cs

```diff
@@ -0,0 +1,175 @@
+using System;
+using Microsoft.SPOT;
+using Microsoft.SPOT.Input;
+using Microsoft.SPOT.Hardware;
+using Microsoft.SPOT.Presentation;
+
+namespace Clock
+{
+    /// <summary>
+    /// Uses the hardware provider to get the pins for handling button input.
+    /// </summary>
+    public sealed class GPIOButtonInputProvider
+    {
+        public readonly Dispatcher Dispatcher;
+
+        private ButtonPad[] buttons;
+        private DispatcherOperationCallback callback;
+        private InputProviderSite site;
+        private PresentationSource source;
+
+        /// <summary>
+        /// Maps GPIOs to Buttons that can be processed by 
+        /// Microsoft.SPOT.Presentation.
+        /// </summary>
+        /// <param name="source"></param>
+        public GPIOButtonInputProvider(PresentationSource source)
+        {
+            // Set the input source.
+            this.source = source;
+
+            // Register our object as an input source with the input manager and 
+            // get back an InputProviderSite object which forwards the input 
+            // report to the input manager, which then places the input in the 
+            // staging area.
+            site = InputManager.CurrentInputManager.RegisterInputProvider(this);
+
+            // Create a delegate that refers to the InputProviderSite object's 
+            // ReportInput method.
+            callback = new DispatcherOperationCallback(delegate(object report)
+                {
+#if MF_FRAMEWORK_VERSION_V3_0

```

Comment:
This is all generated.  I've updated this instance, but we should make the same changes to the template.